validator/cycle/process_tasks.py

```diff
@@ -367,12 +367,47 @@ async def move_tasks_to_preevaluation_loop(config: Config):
 
 
 async def evaluate_tasks_loop(config: Config):
+    gpu_ids = [i for i in range(len(cst.GPU_IDS))]
+
+    # Create queues for tasks and GPUs
+    task_queue = asyncio.Queue()
+    gpu_queue = asyncio.Queue()
+    for gpu_id in gpu_ids:
+        await gpu_queue.put(gpu_id)
+
+    async def evaluation_worker():
```

Comment:
Why have this? seems like we just want to run an eval task on a gpu? This feels very complicated solving a problem we don't have? 

Just async tasks across gpus no?

Reply:
True it looks very complicated at first glance lol

The reason I thought we wouldn't use functions batch by batch is I think it's suboptimal in that the next batch will only start if every single gpu in the current batch completes its docker eval, however tasks can have very different timings because some will have larger models/datasets/number of miners. So the bottleneck will be the biggest task, while all the others that finished quickly will sit idle waiting for it. 

While if we do it dynamically, if just a single gpu is released before the others, it will immediately be put to work again (instead of waiting for all the others)

Another solution I've found is to use semaphores (generalized lock) with the number of gpus, plz lmk if we should discuss online or just implement batching (btw I tested it and I could see the 2 GPUs fire for 2 different tasks, I just need an additional test with bigger tasks)

Reply:
Agree with what you're thinking, but rn the chance of having more than 4 tasks for eval is very small, it's ok there's a bottleneck for simplicity of code - until we have the nice problem of needing to improve this